controllers/slice/namespaces.go

```diff
@@ -0,0 +1,497 @@
+	}
+	err := r.List(ctx, labeledNsList, listOpts...)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Error(err, "Failed to list namespaces")
+			return err
+		}
+	}
+
+	// Sanitize the list of labeled namespaces to guard against someone manually editing the labels
+	// attached to a namespace. The label key and value must match our syntax.
+	// Also build a map for fast lookups in the rest of this function.
+	labeledNsMap := make(map[string]string)
+	for _, labeledNsObj := range labeledNsList.Items {
+		labels := labeledNsObj.ObjectMeta.GetLabels()
+		if labels[AllowedNamespaceSelectorLabelKey] != labeledNsObj.ObjectMeta.Name {
+			log.Error(err, "Incorrect label", "Namespace", labeledNsObj.ObjectMeta.Name)
+			labels[AllowedNamespaceSelectorLabelKey] = labeledNsObj.ObjectMeta.Name
+			labeledNsObj.ObjectMeta.SetLabels(labels)
+			err := r.Update(ctx, &labeledNsObj)
+			if err != nil {
+				log.Error(err, "Failed to update allowed ns label", "Namespace", labeledNsObj.ObjectMeta.Name)
+				return err
+			}
+		}
+		labeledNsMap[labeledNsObj.ObjectMeta.Name] = labels[AllowedNamespaceSelectorLabelKey]
+	}
+
+	// Label the namespace if needed
+	for _, cfgAllowedNs := range cfgAllowedNsList {
+		label, exists := labeledNsMap[cfgAllowedNs]
+		if exists && label == cfgAllowedNs {
+			continue
+		}
+
+		namespace := &corev1.Namespace{}
+		err := r.Get(ctx, types.NamespacedName{Name: cfgAllowedNs}, namespace)
+		if err != nil {
+			log.Info("Failed to find namespace", "namespace", cfgAllowedNs)
+			continue
+		}
+
+		// A namespace might not have any labels attached to it. Directly accessing the label map
+		// leads to a crash for such namespaces.
+		// If the label map is nil, create one and use the setter api to attach it to the namespace.
+		nsLabels := namespace.ObjectMeta.GetLabels()
+		if nsLabels == nil {
+			nsLabels = make(map[string]string)
+		}
+
+		nsLabels[AllowedNamespaceSelectorLabelKey] = cfgAllowedNs
+		namespace.ObjectMeta.SetLabels(nsLabels)
+		err = r.Update(ctx, namespace)
+		if err != nil {
+			log.Error(err, "Failed to label namespace", "Namespace", cfgAllowedNs)
+			return err
+		}
+		log.Info("Labeled namespace successfully", "namespace", cfgAllowedNs)
+	}
+	return nil
+}
+
+func (r *SliceReconciler) unbindAppNamespace(ctx context.Context, slice *kubeslicev1beta1.Slice, appNs string) error {
+	log := logger.FromContext(ctx).WithValues("type", "appNamespaces")
+	debuglog := log.V(1)
+	namespace := &corev1.Namespace{}
+	err := r.Get(ctx, types.NamespacedName{Name: appNs}, namespace)
+	if err != nil {
+		log.Error(err, "NS unbind: Failed to find namespace", "namespace", appNs)
+		return err
+	}
+
+	nsLabels := namespace.ObjectMeta.GetLabels()
+	_, ok := nsLabels[ApplicationNamespaceSelectorLabelKey]
+	if !ok {
+		debuglog.Info("NS unbind: slice label not found", "namespace", appNs)
+	} else {
+		// TBD: For now, we are just deleting the Avesha label from the namespace resource so that it becomes
```

Comment:
Please remove this TBD since this would be taken care of by the annotation removal code from app deployment specs.